libvast/src/format/csv.cpp

```diff
@@ -424,38 +416,36 @@ caf::error reader::read_impl(size_t max_events, size_t max_slice_size,
                              consumer& callback) {
   VAST_ASSERT(max_events > 0);
   VAST_ASSERT(max_slice_size > 0);
-  auto next_line = [&, start = std::chrono::steady_clock::now()] {
-    auto remaining = start + read_timeout_ - std::chrono::steady_clock::now();
-    if (remaining < std::chrono::steady_clock::duration::zero())
-      return true;
-    if (!builder_ || builder_->rows() == 0) {
-      lines_->next();
-      return false;
-    }
-    return lines_->next_timeout(remaining);
+  auto next_line = [&] {
+    auto timed_out = lines_->next_timeout(read_timeout_);
+    if (timed_out)
+      VAST_DEBUG(this, "reached input timeout at line", lines_->line_number());
+    return timed_out;
   };
   if (!parser_) {
-    lines_->next();
+    bool timed_out = next_line();
+    if (timed_out)
+      return ec::timeout;
     auto p = read_header(lines_->get());
     if (!p)
       return p.error();
     parser_ = *std::move(p);
   }
   auto& p = *parser_;
   size_t produced = 0;
+  auto start = std::chrono::steady_clock::now();
   while (produced < max_events) {
     // EOF check.
     if (lines_->done())
       return finish(callback, make_error(ec::end_of_input, "input exhausted"));
-    bool timeout = next_line();
-    // We must check not only for a timeout but also whether any events were
-    // produced to work around CAF's assumption that sources are always able to
-    // generate events. Once `caf::stream_source` can handle empty batches
-    // gracefully, the second check should be removed.
-    if (timeout && produced > 0) {
-      VAST_DEBUG(this, "reached input timeout at line", lines_->line_number());
-      return finish(callback, ec::timeout);
+    if (batch_timeout_ > decltype(batch_timeout_)::zero()
+        && start + batch_timeout_ < std::chrono::steady_clock::now()) {
+      VAST_DEBUG(this, "reached input time limit");
+      break;
     }
+    bool timed_out = next_line();
+    if (timed_out)
+      return ec::timeout;
```

Comment:
Does it make sense to invert the return value of `next_line()`, such that we can write this instead?

```cpp
if (!next_line())
  return ec::timeout;
```

Reply:
The next person reading that code will probably not have the semantics of `next_line()` present. Making the meaning of the return value explicit by naming it `timed_out` helps imho.

Reply:
I don't have too strong of an opinion on this, so feel free to just mark as resolved.